Turn transform definition into configuration style

bradyoo opened the pull request:
Definitions now must be parseable using Properties.load() after the semicolons have been replaced with new lines.

bradyoo commented:
@gpang

bradyoo commented:
The names for these configurations are kept from the old options configurations which means that they are not very good at the moment.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8662/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

bradyoo commented:
@gpang: I have done some basic testing and testing results look good. The only thing that's left are:

1. Making all tests pass
2. Figuring out good names.

bradyoo commented:
renamed hive.file.count.max -> name.count.max 
and renamed hive.file.size.min -> file.size.min

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8699/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8704/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8708/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8710/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8712/
Test FAILed.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8726/
Test FAILed.

bradyoo commented:
Tests in error: 
  GetBlockInfoCommandIntegrationTest.alluxio.client.cli.fsadmin.command.GetBlockInfoCommandIntegrationTest » Runtime

bradyoo commented:
alluxio-bot, test this please.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8727/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8728/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8729/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8730/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8732/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8734/
Test PASSed.

bradyoo commented:
@gpang: ready for review.

bradyoo commented:
@gpang Ping?

gpang commented on table/server/common/src/main/java/alluxio/table/common/transform/TransformPlan.java:
```diff
@@ -44,23 +42,29 @@
-    final List<TransformAction> actions = definition.getActions();
+    ArrayList<JobConfig> actions = new ArrayList<>();
+    Layout baseLayout = mBaseLayout;
+    boolean deleteSrc = false;
```
What is this for?

bradyoo replied:
It's for the case when transformations need to be chained and therefore, for every transformation after the first, you need to deleteSrc.

gpang commented on table/server/common/src/main/java/alluxio/table/common/transform/action/CompactAction.java:
```diff
@@ -48,42 +44,49 @@
+      final String fileSizeString = properties.getProperty(FILE_SIZE_OPTION);
+
+      if (StringUtils.isEmpty(numFilesString) && StringUtils.isEmpty(fileSizeString)) {
+        return null;
```
is this allowed to return `null` as a return value?

bradyoo replied:
Yes. That means Compact is unnecessary.

gpang commented on table/server/common/src/test/java/alluxio/table/common/transform/TransformDefinitionTest.java:
```diff
@@ -24,44 +25,46 @@
   @Test
   public void parse() {
```
Is there a test that tests for normalization of transformation definitions?

bradyoo commented on job/common/src/main/java/alluxio/job/plan/transform/CompactConfig.java:
```diff
@@ -44,10 +44,6 @@
-  /**
-   * The type of database to write the compacted table to.
-   */
-  private final String mDatabaseType;
```
not necessarily. But it's really odd because it's basically unused.

bradyoo commented on table/server/common/src/main/java/alluxio/table/common/transform/action/TransformActionFactory.java:
```diff
@@ -11,24 +11,28 @@
   /**
-   * @return the name of the action
+   * Returns the order of the transform action. Transform actions with a lower number
+   * will be executed before transform actions with a higher number.
```
current behavior is undefined. Would we prefer to define it?

bradyoo commented on table/server/common/src/main/java/alluxio/table/common/transform/action/TransformActionFactory.java:
```diff
@@ -11,24 +11,28 @@
+   *
+   * @return integer representing order number
    */
-  String getName();
```
name was previously used to determine the name of the function which doesn't have a use anymore. We could do something that warrants reintroduction but nothing yet.

bradyoo commented on table/server/common/src/main/java/alluxio/table/common/transform/action/TransformActionFactory.java:
```diff
@@ -11,24 +11,28 @@
    */
-  String getName();
+  default int getOrder() {
+    return 100;
```
Nothing here because there are no other TransformActions.

bradyoo commented on table/server/common/src/main/java/alluxio/table/common/transform/action/TransformActionRegistry.java:
```diff
@@ -37,19 +39,17 @@ private TransformActionRegistry() {} // prevent instantiation
-          String.format("TransformActionFactory for name '%s' does not exist.", name));
+  public static List<TransformAction> create(Properties definition) {
+    final ArrayList<TransformAction> actions = new ArrayList<>();
+    for (TransformActionFactory factory : FACTORIES) {
```
It may be worth it when FACTORIES is a longer list, but it's just so short right now..

bradyoo commented on table/server/common/src/main/java/alluxio/table/common/transform/action/TransformActionRegistry.java:
```diff
@@ -37,19 +39,17 @@ private TransformActionRegistry() {} // prevent instantiation
+    final ArrayList<TransformAction> actions = new ArrayList<>();
+    for (TransformActionFactory factory : FACTORIES) {
+      final TransformAction transformAction = factory.create(definition);
+      if (transformAction != null) {
```
it's not an error for it be to null. It means that the action is unnecessary.

bradyoo commented:
Addressed Feedback. Note that definition normalization is not very good right now and will probably work on that later since it previously was also not very good.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8776/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8783/
Test PASSed.

gpang commented on table/server/common/src/main/java/alluxio/table/common/transform/action/TransformActionFactory.java:
```diff
@@ -11,24 +11,28 @@
    * @return a new instance of an action
    */
-  TransformAction create(String definition, List<String> args, Map<String, String> options);
+  @Nullable
```
can you add to the javadoc on what a `null` return value is supposed to represent?

gpang commented on table/server/common/src/main/java/alluxio/table/common/transform/action/TransformActionRegistry.java:
```diff
@@ -37,19 +39,17 @@ private TransformActionRegistry() {} // prevent instantiation
-    if (factory == null) {
-      throw new IllegalStateException(
-          String.format("TransformActionFactory for name '%s' does not exist.", name));
+  public static List<TransformAction> create(Properties definition) {
```
Can you update the javadoc with the return description and method description? (why is it a list?)

gpang commented on table/server/common/src/test/resources/META-INF/services/alluxio.table.common.transform.action.TransformActionFactory:
```diff
@@ -0,0 +1,2 @@
+alluxio.table.common.transform.action.EarlyActionFactory
+alluxio.table.common.transform.action.LateActionFactory
```
please add a newline at the end of the file

gpang commented on table/server/common/src/main/java/alluxio/table/common/transform/action/TransformActionRegistry.java:
```diff
@@ -59,14 +59,9 @@ private static void refresh() {
     }
-    LOG.info("Registered Transform actions: " + String.join(",", FACTORIES.keySet()));
+    FACTORIES.sort(Comparator.comparingInt((factory) -> factory.getOrder()));
+    LOG.info("Registered Transform actions: " + StringUtils.join(FACTORIES, ","));
```
Can you comment on what this would look like?

bradyoo replied:
Registered Transform actions: alluxio.table.common.transform.action.EarlyActionFactory@4909b8da,CompactActionFactory,alluxio.table.common.transform.action.LateActionFactory@3a03464

CompactActionFactory has a toString implementation, the other 2 don't.

bradyoo commented:
Feedback addressed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8787/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8788/
Test FAILed.

bradyoo commented:
Tests in error: 
  FileSystemAclIntegrationTest.alluxio.client.hadoop.FileSystemAclIntegrationTest » Runtime
Flaked tests: 
alluxio.server.ft.journal.ufs.UfsJournalIntegrationTest.loadMetadata(alluxio.server.ft.journal.ufs.UfsJournalIntegrationTest)
  Run 1: UfsJournalIntegrationTest.loadMetadata:172->loadMetadataTestUtil:176->createFsMasterFromJournal:603 » Runtime
  Run 2: PASS

bradyoo commented:
alluxio-bot, test this please.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8789/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8790/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/8791/
Test PASSed.

bradyoo commented:
Lets merge this @gpang

gpang commented:
alluxio-bot, merge this please
